File: examples/c/validation_custom/validation_custom.c
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */

/**
 * @file validation_custom.c
 *
 * This file contains an example demonstrating how to implement additional
 * Message validation logic in addition to that which the GMSEC API provides.
 */

#include <gmsec4_c.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef _WIN32
	#include <windows.h>

	#define snprintf sprintf_s
#endif


const char* HB_MSG_SUBJECT     = "GMSEC.MISSION.SATELLITE.MSG.HB.VALIDATION-CUSTOM";
const char* ALT_HB_MSG_SUBJECT = "GMSEC.MISSION.SATELLITE.MSG.C2CX.VALIDATION-CUSTOM.HB";


//o Helper functions
void checkStatus(GMSEC_Status status);

void initializeLogging(GMSEC_Config config, GMSEC_Status status);
void setupStandardFields(unsigned int specVersion);
void clearStandardFields();


void customMessageValidator(const GMSEC_Message msg, GMSEC_Status status)
{
	//o Get message type and subtype
	const char* type;
	const char* subtype;

	if (messageHasField(msg, "MESSAGE-TYPE", NULL) == GMSEC_TRUE)
	{
		type = messageGetStringValue(msg, "MESSAGE-TYPE", NULL);
	}
	if (messageHasField(msg, "C2CX-SUBTYPE", NULL) == GMSEC_TRUE)
	{
		subtype = messageGetStringValue(msg, "C2CX-SUBTYPE", NULL);
	}
	else if (messageHasField(msg, "MESSAGE-SUBTYPE", NULL) == GMSEC_TRUE)
	{
		subtype = messageGetStringValue(msg, "MESSAGE-SUBTYPE", NULL);
	}

	//o Ensure we have a Heartbeat message and it contains the PUB-RATE field
	if ((strncmp(type, "MSG", strlen("MSG")) == 0) &&
	    (strncmp(subtype, "HB", strlen("HB")) == 0) &&
	    (messageHasField(msg, "PUB-RATE", NULL) == GMSEC_TRUE))
	{
		GMSEC_I64 pubRate = messageGetIntegerValue(msg, "PUB-RATE", status);

		if (statusIsError(status) == GMSEC_FALSE && ((pubRate < 10) || (pubRate > 60)))
		{
			statusSet(status, GMSEC_API_MSG_ERROR, GMSEC_API_VALUE_OUT_OF_RANGE, "PUB-RATE field does not have a valid value", 0);
		}
	}
	else
	{
		statusSet(status, GMSEC_API_MSG_ERROR, GMSEC_API_INVALID_MSG, "Non-Heartbeat message received", 0);
	}
}


int main(int argc, char* argv[])
{
	GMSEC_Status        status  = NULL;
	GMSEC_Config        config  = NULL;
	GMSEC_ConnectionMgr connMgr = NULL;
	GMSEC_Message       msg     = NULL;
	unsigned int        specVersion;

	if (argc <= 1)
	{
		printf("usage: %s mw-id=<middleware ID>\n", argv[0]);
		return -1;
	}

	status = statusCreate();
	config = configCreateWithArgs(argc, argv);

	initializeLogging(config, status);

	//o Enable Message validation via connection configuration
	configAddValue(config, "gmsec-msg-content-validate-send", "true", NULL);

	GMSEC_INFO("API version: %s", connectionManagerGetAPIVersion());

	connMgr = connectionManagerCreate(config, status);
	checkStatus(status);

	connectionManagerInitialize(connMgr, status);
	checkStatus(status);

	GMSEC_INFO("Middleware version: %s", connectionManagerGetLibraryVersion(connMgr, NULL));

	//o Register custom message validator
	connectionManagerRegisterMessageValidator(connMgr, customMessageValidator, NULL);

	//o Get the version of the message specification we are using
	specVersion = specificationGetVersion(connectionManagerGetSpecification(connMgr, NULL), NULL);

	//o Set up standard/common fields used with all messages
	setupStandardFields(specVersion);

	//o Create Heartbeat Message
	//o Note: Message subject and schema ID vary depending on the specification in use
	const char* subject  = (specVersion > C_GMSEC_ISD_2018_00 ? HB_MSG_SUBJECT : ALT_HB_MSG_SUBJECT);
	const char* schemaID = (specVersion > C_GMSEC_ISD_2018_00 ? "MSG.HB" : "MSG.C2CX.HB");

	msg = mistMessageCreate(subject, schemaID, connectionManagerGetSpecification(connMgr, NULL), status);
	checkStatus(status);

	//o Add PUB-RATE field with illegal value
	mistMessageSetValue(msg, "PUB-RATE", "5", status);
	checkStatus(status);

	//o For very old specifications, we need to add a MSG-ID field
	if (specVersion <= C_GMSEC_ISD_2014_00)
	{
		messageAddStringField(msg, "MSG-ID", "12345", status);
		checkStatus(status);
	}

	//o Attempt to publish malformed message
	GMSEC_INFO("Attempting to publish malformed message...");
	connectionManagerPublish(connMgr, msg, status);

	if (statusIsError(status) == GMSEC_TRUE)
	{
		GMSEC_INFO("This is an expected error:\n%s", statusGet(status));
	}
	else
	{
		GMSEC_WARNING("Was expecting an error");
	}

	//o Fix PUB-RATE field with legal value
	mistMessageSetValue(msg, "PUB-RATE", "15", status);
	checkStatus(status);

	//o Publish a good message
	GMSEC_INFO("Attempting to publish a good message...");
	connectionManagerPublish(connMgr, msg, status);

	if (statusIsError(status) == GMSEC_TRUE)
	{
		GMSEC_WARNING("Unexpected error:\n%s", statusGet(status));
	}
	else
	{
		GMSEC_INFO("Message published!");
	}

	//o Disconnect from the middleware and clean up the Connection
	connectionManagerCleanup(connMgr, status);
	checkStatus(status);

	//o Clear standard/common fields used with all messages
	clearStandardFields();

	messageDestroy(&msg);
	connectionManagerDestroy(&connMgr);
	configDestroy(&config);
	statusDestroy(&status);

	return 0;
}


void checkStatus(GMSEC_Status status)
{
	if (statusIsError(status) == GMSEC_TRUE)
	{
		GMSEC_ERROR("Fatal Error: %s", statusGet(status));
		exit(-1);
	}
}


void initializeLogging(GMSEC_Config config, GMSEC_Status status)
{
	const char* logLevel = configGetValue(config, "LOGLEVEL", status);
	const char* logFile  = configGetValue(config, "LOGFILE", status);

	if (!logLevel)
	{
		configAddValue(config, "LOGLEVEL", "INFO", status);
	}
	if (!logFile)
	{
		configAddValue(config, "LOGFILE", "STDOUT", status);
	}
}


void setupStandardFields(unsigned int specVersion)
{
	size_t numFields = (specVersion <= C_GMSEC_ISD_2016_00 ? 6 : 8);
	size_t i;

	GMSEC_Field* definedFields = malloc(sizeof(GMSEC_Field) * numFields);

	definedFields[0] = stringFieldCreate("MISSION-ID", "MISSION", NULL);
	definedFields[1] = stringFieldCreate("CONSTELLATION-ID", "CONSTELLATION", NULL);
	definedFields[2] = stringFieldCreate("SAT-ID-PHYSICAL", "SATELLITE", NULL);
	definedFields[3] = stringFieldCreate("SAT-ID-LOGICAL", "SATELLITE", NULL);
	definedFields[4] = stringFieldCreate("FACILITY", "GMSEC-LAB", NULL);
	definedFields[5] = stringFieldCreate("COMPONENT", "VALIDATION-CUSTOM", NULL);

	if (specVersion >= C_GMSEC_ISD_2018_00)
	{
		definedFields[6] = stringFieldCreate("DOMAIN1", "DOMAIN1", NULL);
		definedFields[7] = stringFieldCreate("DOMAIN2", "DOMAIN2", NULL);
	}

	mistMessageSetStandardFields(definedFields, numFields, NULL);

	for (i = 0; i < numFields; ++i)
	{
		fieldDestroy(&definedFields[i]);
	}

	free(definedFields);
}


void clearStandardFields()
{
	mistMessageClearStandardFields();
}

File: examples/csharp/validation_custom/validation_custom.cs
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */

/**
 * @file validation_custom.cs
 *
 * This file contains an example demonstrating how to implement additional
 * Message validation logic in addition to that which the GMSEC API provides.
 */

using GMSEC.API;
using System;
using System.Collections.Generic;


class CustomMessageValidator : MessageValidator
{
	public override Status ValidateMessage(Message msg)
	{
		Status status = new Status();

		try
		{
			//o Get message type and subtype
			string type    = "";
			string subtype = "";

			if (msg.HasField("MESSAGE-TYPE"))
			{
				type = msg.GetStringValue("MESSAGE-TYPE");
			}
			if (msg.HasField("C2CX-SUBTYPE"))
			{
				subtype = msg.GetStringValue("C2CX-SUBTYPE");
			}
			else if (msg.HasField("MESSAGE-SUBTYPE"))
			{
				subtype = msg.GetStringValue("MESSAGE-SUBTYPE");
			}

			//o Ensure we have a Heartbeat message and it contains the PUB-RATE field
			if (type == "MSG" && subtype == "HB" && msg.HasField("PUB-RATE"))
			{
				Int64 pubRate = msg.GetIntegerValue("PUB-RATE");

				//o Ensure PUB-RATE field value is between 10 and 60 (inclusive)
				if (pubRate < 10 || pubRate > 60)
				{
					status.Set(StatusClass.MSG_ERROR, StatusCode.VALUE_OUT_OF_RANGE, "PUB-RATE field does not have a valid value");
				}
			}
			else
			{
				status.Set(StatusClass.MSG_ERROR, StatusCode.INVALID_MSG, "Non-Heartbeat message received");
			}
		}
		catch (GmsecException e)
		{
			status = new Status(e.GetErrorClass(), e.GetErrorCode(), e.GetErrorMessage(), e.GetCustomCode());
		}

		return status;
	}
}


class validation_custom
{
	static private string HB_MSG_SUBJECT     = "GMSEC.MISSION.SATELLITE.MSG.HB.VALIDATION-CUSTOM";
	static private string ALT_HB_MSG_SUBJECT = "GMSEC.MISSION.SATELLITE.MSG.C2CX.VALIDATION-CUSTOM.HB";


	static int Main(string[] args)
	{
		if (args.Length < 1)
		{
			Console.WriteLine("usage validation_custom.exe mw-id=<middleware ID>");
			return -1;
		}

		Config config = new Config(args);

		InitializeLogging(config);

		//o Enable Message validation.  This parameter is "false" by default.
		config.AddValue("gmsec-msg-content-validate-send", "true");
		config.AddValue("gmsec-validation-level", "3");

		Log.Info("API version: " + ConnectionManager.GetAPIVersion());

		try
		{
			ConnectionManager connMgr = new ConnectionManager(config);

			connMgr.Initialize();

			Log.Info("Middleware version: " + connMgr.GetLibraryVersion());

			//o Register custom message validator
			CustomMessageValidator cmv = new CustomMessageValidator();
			connMgr.RegisterMessageValidator(cmv);

			//o Create and publish a simple Product File Message
			uint specVersion = connMgr.GetSpecification().GetVersion();
			SetupStandardFields(specVersion);

			//o Create Heartbeat Message
			//o Note: Message subject and schema ID vary depending on the specification in use
			string subject  = (specVersion > Gmsec.GMSEC_ISD_2018_00 ? HB_MSG_SUBJECT : ALT_HB_MSG_SUBJECT);
			string schemaID = (specVersion > Gmsec.GMSEC_ISD_2018_00 ? "MSG.HB" : "MSG.C2CX.HB");

			MistMessage msg = new MistMessage(subject, schemaID, connMgr.GetSpecification());

			//o Add PUB-RATE field with illegal value
			msg.SetValue("PUB-RATE", "5");

			//o For very old specifications, we need to add a MSG-ID field
			if (specVersion <= Gmsec.GMSEC_ISD_2014_00)
			{
				msg.AddField("MSG-ID", "12345");
			}

			//o Attempt to publish malformed message
			try
			{
				Log.Info("Attempting to publish malformed message...");
				connMgr.Publish(msg);

				Log.Warning("Was expecting an error");
			}
			catch (GmsecException e)
			{
				//o We expect to encounter error with the PUB-RATE field
				Log.Info("This is an expected error:\n" + e.ToString());
			}

			//o Fix PUB-RATE field with legal value
			msg.SetValue("PUB-RATE", "15");

			//o Publish a good message
			try
			{
				Log.Info("Attempting to publish good message...");
				connMgr.Publish(msg);

				Log.Info("Message published!");
			}
			catch (GmsecException e)
			{
				Log.Warning("Unexpected error:\n" + e.ToString());
			}

			//o Disconnect from the middleware and clean up the Connection
			connMgr.Cleanup();

			//o Clear standard/common fields used with all messages
			ClearStandardFields();
		}
		catch (GmsecException e)
		{
			Log.Error(e.ToString());
			return -1;
		}

		return 0;
	}

	static void InitializeLogging(Config config)
	{
		String logLevel = config.GetValue("LOGLEVEL");
		String logFile = config.GetValue("LOGFILE");

		if (logLevel == null)
		{
			config.AddValue("LOGLEVEL", "INFO");
		}
		if (logFile == null)
		{
			config.AddValue("LOGFILE", "STDERR");
		}
	}

	static void SetupStandardFields(uint specVersion)
	{
		FieldList definedFields = new FieldList();

		StringField mission       = new StringField("MISSION-ID", "MISSION");
		StringField constellation = new StringField("CONSTELLATION-ID", "CONSTELLATION");
		StringField satIdPhys     = new StringField("SAT-ID-PHYSICAL", "SATELLITE");
		StringField satIdLog      = new StringField("SAT-ID-LOGICAL", "SATELLITE");
		StringField facility      = new StringField("FACILITY", "GMSEC-LAB");
		StringField component     = new StringField("COMPONENT", "VALIDATE-CUSTOM");

		definedFields.Add(mission);
		definedFields.Add(constellation);
		definedFields.Add(satIdPhys);
		definedFields.Add(satIdLog);
		definedFields.Add(facility);
		definedFields.Add(component);

		if (specVersion >= Gmsec.GMSEC_ISD_2018_00)
		{
			StringField domain1 = new StringField("DOMAIN1", "DOMAIN1");
			StringField domain2 = new StringField("DOMAIN2", "DOMAIN2");

			definedFields.Add(domain1);
			definedFields.Add(domain2);
		}

		MistMessage.SetStandardFields(definedFields);
	}

	static void ClearStandardFields()
	{
		MistMessage.ClearStandardFields();
	}
}
